core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -301,9 +301,15 @@ private[kafka] object LogValidator extends Logging {
       throw new InvalidRecordException("Control messages are not currently supported")
   }
 
-  private def validateKey(record: Record, compactedTopic: Boolean) {
-    if (compactedTopic && !record.hasKey)
-      throw new InvalidRecordException("Compacted topic cannot accept message without key.")
+  private def validateKey(record: Record, compactedTopic: Boolean, recordBatch: RecordBatch) {
+    if (compactedTopic && !record.hasKey) {
+      // throw a non retriable exception only if the produce request is V3 or higher
+      if (record.hasMagic(recordBatch.magic)) {
```

Comment:
This check does not look correct. It seems like it may be easiest to do the conditional logic in a higher layer where we know the produce request version. For example, it's possible for message format V0 to be used with produce request V3.